x/stake/keeper/delegation.go

```diff
@@ -24,41 +24,164 @@ func (k Keeper) GetDelegation(ctx sdk.Context,
+	}
+	store := ctx.KVStore(k.storeKey)
+	delegatorPrefixKey := GetDelegationsKey(delegatorAddr)
+	iterator := sdk.KVStorePrefixIterator(store, delegatorPrefixKey) //smallest to largest
+
+	i := 0
+	for ; iterator.Valid() && (!retrieve || (retrieve && i < int(maxRetrieve[0]))); iterator.Next() {
+		addr := iterator.Key()
+		delegation := types.MustUnmarshalDelegation(k.cdc, addr, iterator.Value())
+		validator, found := k.GetValidator(ctx, delegation.ValidatorAddr)
+		if !found {
+			panic(types.ErrNoValidatorFound(types.DefaultCodespace))
+		}
+
+		if retrieve {
+			validators[i] = validator
+		} else {
+			validators = append(validators, validator)
+		}
+
+		i++
+	}
+	iterator.Close()
+	return validators[:i] // trim
+}
+
+// return a validator that a delegator is bonded to
+func (k Keeper) GetDelegatorBechValidator(ctx sdk.Context, delegatorAddr sdk.AccAddress,
+	validatorAddr sdk.ValAddress) (bechValidator types.BechValidator) {
+
+	validator := k.GetDelegatorValidator(ctx, delegatorAddr, validatorAddr)
+	bechValidator, err := validator.Bech32Validator()
+	if err != nil {
+		panic(err.Error())
+	}
+
+	return
+}
+
+// return a validator that a delegator is bonded to
+func (k Keeper) GetDelegatorValidator(ctx sdk.Context, delegatorAddr sdk.AccAddress,
+	validatorAddr sdk.ValAddress) (validator types.Validator) {
+
+	delegation, found := k.GetDelegation(ctx, delegatorAddr, validatorAddr)
+	if !found {
+		panic(types.ErrNoDelegation(types.DefaultCodespace))
+	}
+
+	validator, found = k.GetValidator(ctx, delegation.ValidatorAddr)
+	if !found {
+		panic(types.ErrNoValidatorFound(types.DefaultCodespace))
+	}
+	return
+}
 
+// Return all delegations for a delegator. If maxRetrieve is supplied, the respective amount will be returned.
+func (k Keeper) GetDelegatorDelegations(ctx sdk.Context, delegator sdk.AccAddress,
+	maxRetrieve ...int16) (delegations []types.Delegation) {
+	retrieve := len(maxRetrieve) > 0
+	if retrieve {
+		delegations = make([]types.Delegation, maxRetrieve[0])
+	}
 	store := ctx.KVStore(k.storeKey)
 	delegatorPrefixKey := GetDelegationsKey(delegator)
 	iterator := sdk.KVStorePrefixIterator(store, delegatorPrefixKey) //smallest to largest
 
-	delegations = make([]types.Delegation, maxRetrieve)
 	i := 0
-	for ; ; i++ {
-		if !iterator.Valid() || i > int(maxRetrieve-1) {
-			break
+	for ; iterator.Valid() && (!retrieve || (retrieve && i < int(maxRetrieve[0]))); iterator.Next() {
+		delegation := types.MustUnmarshalDelegation(k.cdc, iterator.Key(), iterator.Value())
+		if retrieve {
+			delegations[i] = delegation
+		} else {
+			delegations = append(delegations, delegation)
 		}
+		i++
+	}
+	iterator.Close()
+	return delegations[:i] // trim
+}
+
+// Return all delegations for a delegator. If maxRetrieve is supplied, the respective amount will be returned.
+func (k Keeper) GetDelegatorDelegationsWithoutRat(ctx sdk.Context, delegator sdk.AccAddress,
+	maxRetrieve ...int16) (delegations []types.DelegationWithoutDec) {
```

Comment:
Remove this function altogether - the `DelegationForREST` object should be a property of the `Delegation` object and should be retrieved after collecting the delegations with `GetDelegations` - functions like this bloat the keeper